### PR TITLE
Allow non-strict Gradle classloading to support evaluationDependsOn() & COD

### PR DESCRIPTION
**Pull request by eskatos**
This should address both #311 and #607

### CODE REVIEW

File: provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/GradleKotlinDslIntegrationTest.kt [hunk 1]
```diff
@@ -894,6 +896,28 @@ class GradleKotlinDslIntegrationTest : AbstractIntegrationTest() {
...
     }
 
+    /**
+     * Assert that the Kotlin DSL provider don't force strict classloading
+     * by exercising a case where :b class loader scopes aren't locked when evaluating :b:c.
+     */
+    @Test
+    fun `allow for non-strict Gradle classloading`() {
+
+        val previous = System.getProperty(DefaultClassLoaderScope.STRICT_MODE_PROPERTY)
```
Comment (bamboo) [inline provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/GradleKotlinDslIntegrationTest.kt]: :x: `withSystemProperty(STRICT_MODE_PROPERTY, "false") { ...`

**Comment by oehme** [comment]
I agree we should fix this upstream. The Kotlin DSL shouldn't make that decision though, which is why I'm 👍 on this change.

**Comment by bamboo** [comment]
I understand where you're coming from @oehme but I still don't think the decision the Kotlin DSL is making here - `require(scope.isLocked)` - is unreasonable at this level.

It's there to enforce the reproducibility of computing the classpath of a project.

The alternative, to let the classpath vary depending on the order the projects are evaluated, could lead to hard to debug script compilation errors particularly when combined with configure-on-demand:

> Why my script fails to compile when I run `./gradlew :foo:bar` but compiles cleanly when I run `./gradlew bar`?

Or am I overstating the risk?

**Comment by oehme** [comment]
The same is true for Groovy scripts and has been for a long time. That's why I don't think this check belongs in the Kotlin DSL. We should fix this centrally.

**Comment by eskatos** [comment]
> The same is true for Groovy scripts and has been for a long time. That's why I don't think this check belongs in the Kotlin DSL. We should fix this centrally.

@oehme could you please express your concern an an issue on `gradle/gradle` so we can refer to it if need be? or there is one I couldn't find already?

**Comment by eskatos** [comment]
Closing as superseded by https://github.com/gradle/gradle/issues/4823

**Pull request closed without merging**